回归树

数据

|x|1|2|3|4|5|6|7|8|9|10
|--|--|--|--|--|--|--|--|--|--|--|
|y|4.50|4.75|4.91|5.34|5.80|7.05|7.90|8.23|8.70|9.00|

根据 x 的取值将数据划分为两组

例如划分点为 1.5，那么数据将被划分为两组，一组是 {1}, 另一组为 {2,3,4,...,10}

将每一组数据 y 的平均值作为这一组的预测值

那么第一组的预测值为 4.50

第二组预测值为 

In [2]:
1/9 *(4.75+4.91+5.34+5.80+7.05+7.90+8.23+8.70+9.00)

6.8533333333333335

|x|1|2|3|4|5|6|7|8|9|10
|--|--|--|--|--|--|--|--|--|--|--|
|y|4.50|4.75|4.91|5.34|5.80|7.05|7.90|8.23|8.70|9.00|
|预测值|4.50|6.85|6.85|6.85|6.85|6.85|6.85|6.85|6.85|6.85|

然后计算一下均方误差：(y - 预测值)^2

In [5]:
import numpy as np

In [14]:
x = np.arange(1,11)
y = np.array([4.50,4.75,4.91,5.34,5.80,7.05,7.90,8.23,8.70,9.00])

In [21]:
predict = np.array([4.50,6.85,6.85,6.85,6.85,6.85,6.85,6.85,6.85,6.85])

In [28]:
(predict - y)**2

array([0.    , 4.41  , 3.7636, 2.2801, 1.1025, 0.04  , 1.1025, 1.9044,
       3.4225, 4.6225])

|x|1|2|3|4|5|6|7|8|9|10
|--|--|--|--|--|--|--|--|--|--|--|
|y|4.50|4.75|4.91|5.34|5.80|7.05|7.90|8.23|8.70|9.00|
|预测值|4.50|6.85|6.85|6.85|6.85|6.85|6.85|6.85|6.85|6.85|
|mse|0.    | 4.41  | 3.7636| 2.2801| 1.1025| 0.04  | 1.1025| 1.9044|3.4225| 4.6225

对 mse 取一下和，作为这次划分的评价

In [45]:
sum((predict - y)**2)

17.702399999999997

即以 1.5 为划分点，预测误差为 22.6481

接下来以 2.5 为划分点，重复以上步骤

以2.5为划分点，数据被分为两组 {1,2}, {3,4,5,6,7,8,9,10}

第一组的预测值：

In [46]:
sum(y[:2])/2

4.625

第二组的预测值：

In [47]:
sum(y[2:])/8

7.116250000000001

|x|1|2|3|4|5|6|7|8|9|10
|--|--|--|--|--|--|--|--|--|--|--|
|y|4.50|4.75|4.91|5.34|5.80|7.05|7.90|8.23|8.70|9.00|
|预测值|4.63|4.63|7.12|7.12|7.12|7.12|7.12|7.12|7.12|7.12|
|mse|0.    | 4.41  | 3.7636| 2.2801| 1.1025| 0.04  | 1.1025| 1.9044|3.4225| 4.6225

In [37]:
predict = np.array([4.63]*2 + [7.12]*8)
predict

array([4.63, 4.63, 7.12, 7.12, 7.12, 7.12, 7.12, 7.12, 7.12, 7.12])

计算一下 mse

In [38]:
(predict - y)**2

array([0.0169, 0.0144, 4.8841, 3.1684, 1.7424, 0.0049, 0.6084, 1.2321,
       2.4964, 3.5344])

|x|1|2|3|4|5|6|7|8|9|10
|--|--|--|--|--|--|--|--|--|--|--|
|y|4.50|4.75|4.91|5.34|5.80|7.05|7.90|8.23|8.70|9.00|
|预测值|4.63|4.63|7.12|7.12|7.12|7.12|7.12|7.12|7.12|7.12|
|mse|0.0169| 0.0144| 4.8841| 3.1684| 1.7424| 0.0049| 0.6084| 1.2321|2.4964| 3.5344|3.4225| 4.6225

那么以 2.5 为划分点的预测误差为：

In [41]:
sum((predict - y)**2)

17.702399999999997

接下来我们继续尝试以 2.5, 3.5, 4.5, ..., 9.5 为划分点，分别计算它们的预测误差

|划分点|1.5|2.5|3.5|4.5|5.5|6.5|7.5|8.5|9.5|
|--|--|--|--|--|--|--|--|--|--|
|预测误差|22.65| 17.70| 12.19| 7.38| 3.36| 5.07| 10.05| 15.18|21.33|

当划分点为 5.5 时，预测误差最小，所以选择 5.5 为最优划分点

以 5.5 为划分点将数据集分为两个部分
{1,2,3,4,5} 和 {6,7,8,9,10}

![avatar](test.png)

此时回归树的预测为：

|x|1|2|3|4|5|6|7|8|9|10
|--|--|--|--|--|--|--|--|--|--|--|
|y|4.50|4.75|4.91|5.34|5.80|7.05|7.90|8.23|8.70|9.00|
|预测值|5.06|5.06|5.06|5.06|5.06|8.17|8.17|8.17|8.17|8.17|

8.175999999999998

利用同样的方法分别对这两部分继续划分

对第一个部分，可以选择的划分点有 1.5, 2.5, 3.5, 4.5

|x|1|2|3|4|5
|--|--|--|--|--|--|
|y|4.50|4.75|4.91|5.34|5.80|

分别计算分类误差

|划分点|1.5|2.5|3.5|4.5|
|--|--|--|--|--|
|预测误差|0.67| 0.43| 0.19| 0.37|

划分点为 3.5 时，误差最小，因此 {1,2,3,4,5} 被进一步的二分为 {1,2,3} 和 {4,5}

![avatar](test1.png)

|x|1|2|3|4|5|6|7|8|9|10
|--|--|--|--|--|--|--|--|--|--|--|
|y|4.50|4.75|4.91|5.34|5.80|7.05|7.90|8.23|8.70|9.00|
|预测值|4.72|4.72|4.72|5.57|5.57|8.17|8.17|8.17|8.17|8.17|

In [55]:
sum(y[3:5])/2

5.57

继续递归这个过程，最终生成一棵回归树